File: new/ZXNoterApp/src/team/zxorg/zxnoter/core/TreeNode.java
package team.zxorg.zxnoter.core;

import org.jetbrains.annotations.NotNull;
import team.zxorg.zxnoter.api.node.ITreeNode;

import java.util.*;
import java.util.function.Predicate;

public class TreeNode implements ITreeNode {
    private final String name;
    private final Map<String, ITreeNode> children = new HashMap<>();
    private final ITreeNode parent;
    private Object value;

    protected TreeNode(String name, ITreeNode parent) {
        this.name = sanitizeKey(name);
        this.parent = parent;
    }


    public TreeNode(String name) {
        this.name = name;
        this.parent = null;
    }

    private static String sanitizeKey(String name) {
        return name.toLowerCase(Locale.ROOT).replaceAll("\\.", "");
    }

    @Override
    public <T> T getValue(@NotNull Class<T> tclass) {
        return tclass.cast(value);
    }

    @Override
    public <T> void setValue(@NotNull T value) {
        this.value = value;
    }

    @Override
    public ITreeNode getParentNode() {
        return parent;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public ITreeNode addNode(@NotNull String name) {
        return children.computeIfAbsent(sanitizeKey(name), s -> new TreeNode(s, this));
    }

    @Override
    public ITreeNode getNode(@NotNull String name) {
        return children.get(sanitizeKey(name));
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj instanceof ITreeNode key)
            return Objects.equals(getPath(), key.getPath());
        return false;
    }

    @Override
    public int hashCode() {
        return Objects.hash(getPath());
    }

    @Override
    public String toString() {
        return getPath();
    }

    @Override
    public String getPath() {
        List<String> path = new LinkedList<>();
        ITreeNode current = this;
        while (current != null) {
            path.addFirst(current.getName());
            current = current.getParentNode();
        }
        return String.join(".", path);
    }

    @Override
    public Collection<ITreeNode> getChildren() {
        return Collections.unmodifiableCollection(children.values());
    }

    public ITreeNode findByName(String name) {
        if (name == null) {
            return null;
        }
        // 先检查当前节点的子节点
        ITreeNode node = children.get(sanitizeKey(name));
        if (node != null) {
            return node;
        }
        // 递归地在子节点中查找
        for (ITreeNode child : children.values()) {
            ITreeNode result = ((TreeNode) child).findByName(name);
            if (result != null) {
                return result;
            }
        }
        return null;
    }

    @Override
    public List<ITreeNode> findByCondition(Predicate<String> condition) {
        List<ITreeNode> result = new ArrayList<>();
        if (condition.test(name)) {
            result.add(this);
        }
        for (ITreeNode child : children.values()) {
            result.addAll(child.findByCondition(condition));
        }
        return result;
    }


}

File: utils/src/main/kotlin/org/zxnoter/utils/node/TreeNode.kt
package org.zxnoter.utils.node


/**
 * 树节点
 */
interface TreeNode<T> {
    /**
     * 获取子节点数量
     */
    fun size(): Int

    /**
     * 获取父节点
     * @return 如果当前节点是根节点则返回 null
     */
    fun parent(): TreeNode<T>?

    /**
     * 获取树的根节点
     */
    fun root(): TreeNode<T>

    /**
     * 获取所有子节点
     */
    fun children(): List<TreeNode<T>>

    /**
     * 增加子节点
     * @return 新增的节点
     */
    fun add(name: String, value: T): TreeNode<T>

    /**
     * 获取子节点
     */
    fun get(name: String): TreeNode<T>?

    /**
     * 移除子节点
     * @return 被移除的节点
     */
    fun remove(name: String): TreeNode<T>?

    /***
     * 获取节点值
     */
    val value: T?

    /**
     * 查找路径对应的节点
     */
    fun find(regex: Regex): List<TreeNode<T>>

    /**
     * 当前节点的名称
     */
    val name: String?

    /**
     * 获取当前节点的路径
     */
    fun path(): NodePath

    /**
     * 检查当前节点是否为叶子节点
     */
    fun isLeaf(): Boolean

    /**
     * 获取当前节点的深度
     */
    fun depth(): Int

    /**
     * 添加一个监听器
     */
//    fun addListener(listener: TreeNodeListener<T>)

    /**
     * 移除一个监听器
     */
//    fun removeListener(listener: TreeNodeListener<T>)
}


